File: slang-plugin/src/test/java/org/sonarsource/slang/plugin/MetricVisitorTest.java
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.slang.plugin;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.HashSet;
import java.util.Set;
import org.junit.Rule;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.migrationsupport.rules.EnableRuleMigrationSupport;
import org.junit.rules.TemporaryFolder;
import org.sonar.api.batch.fs.InputFile;
import org.sonar.api.batch.fs.internal.DefaultInputFile;
import org.sonar.api.batch.fs.internal.TestInputFileBuilder;
import org.sonar.api.batch.sensor.internal.SensorContextTester;
import org.sonar.api.issue.NoSonarFilter;
import org.sonar.api.measures.FileLinesContext;
import org.sonar.api.measures.FileLinesContextFactory;
import org.sonarsource.slang.parser.SLangConverter;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@EnableRuleMigrationSupport
class MetricVisitorTest {

  private NoSonarFilter mockNoSonarFilter;
  private SLangConverter parser = new SLangConverter();
  private MetricVisitor visitor;
  private SensorContextTester sensorContext;
  private DefaultInputFile inputFile;

  @Rule
  public TemporaryFolder tempFolder = new TemporaryFolder();

  @BeforeEach
  void setUp() {
    sensorContext = SensorContextTester.create(tempFolder.getRoot());
    FileLinesContext mockFileLinesContext = mock(FileLinesContext.class);
    FileLinesContextFactory mockFileLinesContextFactory = mock(FileLinesContextFactory.class);
    mockNoSonarFilter = mock(NoSonarFilter.class);
    when(mockFileLinesContextFactory.createFor(any(InputFile.class))).thenReturn(mockFileLinesContext);
    visitor = new MetricVisitor(mockFileLinesContextFactory, mockNoSonarFilter);
  }

  @Test
  void emptySource() throws Exception {
    scan("");
    assertThat(visitor.linesOfCode()).isEmpty();
    assertThat(visitor.commentLines()).isEmpty();
    assertThat(visitor.numberOfFunctions()).isZero();
    verify(mockNoSonarFilter).noSonarInFile(inputFile, new HashSet<>());
  }

  @Test
  void linesOfCode() throws Exception {
    scan("" +
      "x + 1;\n" +
      "// comment\n" +
      "fun function1() { // comment\n" +
      "x = true || false; }");
    assertThat(visitor.linesOfCode()).containsExactly(1, 3, 4);
  }

  @Test
  void commentLines() throws Exception {
    scan("" +
      "x + 1;\n" +
      "// comment\n" +
      "fun function1() { // comment\n" +
      "x = true || false; }");
    assertThat(visitor.commentLines()).containsExactly(2, 3);
  }

  @Test
  void multiLineComment() throws Exception {
    scan("" +
      "/*start\n" +
      "x + 1\n" +
      "end*/");
    assertThat(visitor.commentLines()).containsExactly(1, 2, 3);
    assertThat(visitor.linesOfCode()).isEmpty();
  }

  @Test
  void nosonarLines() throws Exception {
    scan("" +
      "x + 1;\n" +
      "// NOSONAR comment\n" +
      "fun function1() { // comment\n" +
      "x = true || false; }");
    assertThat(visitor.nosonarLines()).containsExactly(2);
    Set<Integer> nosonarLines = new HashSet<>();
    nosonarLines.add(2);
    verify(mockNoSonarFilter).noSonarInFile(inputFile, nosonarLines);
  }

  @Test
  void functions() throws Exception {
    scan("" +
      "x + 1;\n" +
      "x = true || false;");
    assertThat(visitor.numberOfFunctions()).isZero();
    scan("" +
      "x + 1;\n" +
      "fun noBodyFunction();\n" + // Only functions with implementation bodies are considered for the metric
      "fun() { x = 1; }\n" + // Anonymous functions are not considered for function metric computation
      "fun function1() { // comment\n" +
      "x = true || false; }");
    assertThat(visitor.numberOfFunctions()).isEqualTo(1);
  }

  @Test
  void classes() throws Exception {
    scan("" +
            "x + 1;\n" +
            "x = true || false;");
    assertThat(visitor.numberOfClasses()).isZero();
    scan("" +
            "class C {}\n" +
            "fun function() {}\n" +
            "class D { int val x = 0; }\n" +
            "class E {\n" +
            "  fun doSomething(int x) {}\n" +
            "}");
    assertThat(visitor.numberOfClasses()).isEqualTo(3);
  }

  @Test
  void cognitiveComplexity() throws Exception {
    scan("" +
      "class A { fun foo() { if(1 != 1) 1; } }" + // +1 for 'if'
      "fun function() {" +
      "  if (1 != 1) {" + // +1 for 'if'
      "    if (1 != 1) {" + // + 2 for nested 'if'
      "      1" +
      "    }" +
      "  };" +
      "  class B {" + // Nesting level reset here because of class declaration
      "    fun bar(int a) {" +
      "      match(a) {" + // +1 for match
      "        1 -> doSomething();" +
      "        2 -> doSomething();" +
      "        else -> if (1 != 1) doSomething();" + // +2 for nested 'if'
      "      }" +
      "    }" +
      "  };" +
      "}");
    assertThat(visitor.cognitiveComplexity()).isEqualTo(7);
  }

  @Test
  void executable_lines() throws Exception {
    scan("" +
      "package abc;\n" +
      "import x;\n" +
      "class A {\n" +
      "  fun foo() {\n" +
      "    statementOnSeveralLines(a,\n" +
      "      b);\n" +
      "  };\n" +
      "}\n" +
      "{\n" +
      "  x = 42;\n" +
      "};");
    assertThat(visitor.executableLines()).containsExactly(5, 10);
  }

  private void scan(String code) throws IOException {
    inputFile = new TestInputFileBuilder("moduleKey", tempFolder.newFile().getName())
      .setCharset(StandardCharsets.UTF_8)
      .initMetadata(code).build();
    InputFileContext ctx = new InputFileContext(sensorContext, inputFile);
    visitor.scan(ctx, parser.parse(code));
  }

}

File: sonar-kotlin-plugin/src/test/java/org/sonarsource/kotlin/plugin/MetricVisitorTest.kt
/*
 * SonarSource Kotlin
 * Copyright (C) 2018-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.kotlin.plugin

import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.jetbrains.kotlin.config.LanguageVersion
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.sonar.api.batch.fs.InputFile
import org.sonar.api.batch.fs.internal.DefaultInputFile
import org.sonar.api.batch.fs.internal.TestInputFileBuilder
import org.sonar.api.batch.sensor.internal.SensorContextTester
import org.sonar.api.issue.NoSonarFilter
import org.sonar.api.measures.FileLinesContext
import org.sonar.api.measures.FileLinesContextFactory
import org.sonarsource.kotlin.converter.Environment
import org.sonarsource.kotlin.converter.KotlinTree
import org.sonarsource.kotlin.utils.kotlinTreeOf
import java.nio.charset.StandardCharsets
import java.nio.file.Path
import kotlin.io.path.createTempFile
import kotlin.io.path.name

internal class MetricVisitorTest {
    private val environment = Environment(emptyList(), LanguageVersion.LATEST_STABLE)
    private lateinit var mockNoSonarFilter: NoSonarFilter
    private lateinit var visitor: MetricVisitor
    private lateinit var sensorContext: SensorContextTester
    private lateinit var inputFile: DefaultInputFile

    @JvmField
    @TempDir
    var tempFolder: Path? = null

    @BeforeEach
    fun setUp() {
        sensorContext = SensorContextTester.create(tempFolder!!.root)
        val mockFileLinesContext = Mockito.mock(FileLinesContext::class.java)
        val mockFileLinesContextFactory = Mockito.mock(
            FileLinesContextFactory::class.java)
        mockNoSonarFilter = Mockito.mock(NoSonarFilter::class.java)
        Mockito.`when`(mockFileLinesContextFactory.createFor(ArgumentMatchers.any(
            InputFile::class.java))).thenReturn(mockFileLinesContext)
        visitor = MetricVisitor(mockFileLinesContextFactory, mockNoSonarFilter)
    }

    @Test
    fun linesOfCode() {
        scan("""
    fun function1(x: Int) { x + 1 }
    // comment
    fun function1() { // comment
        val x = true || false 
    }
    """.trimIndent())
        assertThat(visitor.linesOfCode()).containsExactly(1, 3, 4, 5)
    }

    @Test
    fun commentLines() {
        scan("""
    fun function1(x: Int) { x + 1 }
    // comment
    fun function1() { // comment
        val x = true || false
    }
    """.trimIndent())
        assertThat(visitor.commentLines()).containsExactly(2, 3)
    }

    @Test
    fun multiLineComment() {
        scan("""
    /*start
    x + 1
    end*/
    """.trimIndent())
        assertThat(visitor.commentLines()).containsExactly(1, 2, 3)
        assertThat(visitor.linesOfCode()).isEmpty()
    }

    @Test
    fun nosonarLines() {
        scan("""
    fun function1(x: Int) { x + 1 }
    // NOSONAR comment
    fun function2() { // comment
        val x = true || false
    }
    """.trimIndent())
        assertThat(visitor.nosonarLines()).containsExactly(2)
        assertThat(visitor.commentLines()).containsExactly(3)
        Mockito.verify(mockNoSonarFilter).noSonarInFile(inputFile, setOf(2))
    }

    @Test
    fun `kDocs are counted`() {
        scan("""
    fun function1(x: Int) { x + 1 }
    /**
     * KDoc comment
     */
    fun function2() {
        val x = true || false
    }
    """.trimIndent())
        assertThat(visitor.commentLines()).containsExactly(2, 3, 4)
        assertThat(visitor.nosonarLines()).isEmpty()
    }

    @Test
    fun `header comments are not counted`() {
        scan("""
    // Header comment
    //
    package a
    
    fun function1(x: Int) { x + 1 }
    """.trimIndent())
        assertThat(visitor.commentLines()).isEmpty()
    }


    @Test
    fun `header block comments are not counted`() {
        scan("""
            
    /*
     * Header comment
     */
    package b
      
    val my_c = 2
     
    fun function1(x: Int) { x + 1 }
    """.trimIndent())
        assertThat(visitor.commentLines()).isEmpty()
    }

    @Test
    fun `double header comments are counted correctly`() {
        scan("""
            
    /*
     * Header comment
     */
     
    // Another header comment
     
    package b
      
    val my_c = 2
     
    fun function1(x: Int) { x + 1 }
    """.trimIndent())
        assertThat(visitor.commentLines()).isEmpty()
    }

    @Test
    fun `a mixture of comments is counted correctly`() {
        scan("""
            
    /*
     * Header comment
     */
    package b
      
    // This is a comment
    // for the variable my_c
    val my_c = 2
     
    // NOSONAR comment
    
    /**
     * A KDoc comment
     */
    fun function1(x: Int) { x + 1 } // A comment

    // NOSONAR comment
    """.trimIndent())
        assertThat(visitor.commentLines()).containsExactly(7, 8, 13, 14, 15, 16)
        assertThat(visitor.nosonarLines()).containsExactly(11, 18)
    }


    @Test
    fun functions() {
        scan("""
            class C {
                init {
                    val y = 1 + 1;
                    val x = true || false;
                }
            }
    """.trimIndent())
        assertThat(visitor.numberOfFunctions()).isZero

        scan("""
            abstract class C {
                fun noBodyFunction()
                fun function1() { // comment
                    val x = true || false
                }
            }
    """.trimIndent())
        assertThat(visitor.numberOfFunctions()).isEqualTo(1)
    }

    @Test
    fun classes() {
        scan("""
             fun foo(x: Int) { x + 1 }
             fun bar() = true || false;
    """.trimIndent())
        assertThat(visitor.numberOfClasses()).isZero

        scan("""class C {}
                fun function() {}
                class D { val x = 0 }
                class E {
                  fun doSomething(x: Int) {}
                }""")
        assertThat(visitor.numberOfClasses()).isEqualTo(3)
    }

    @Test
    fun cognitiveComplexity() {
        scan("""
                class A { fun foo() { if(1 != 1) return 1 } } // +1
                    fun function() {
                      if (1 != 1) { // +1
                        if (1 != 1) { // +2
                          1
                        }
                      }
                      class B {
                        fun bar(a: Int) {
                          when(a) { // +1
                            1 -> doSomething()
                            2 -> doSomething()
                            else -> if (1 != 1) doSomething(); // +2
                          }
                        }
                      }
                }""")
        assertThat(visitor.cognitiveComplexity()).isEqualTo(7)
    }

    @Test
    fun executable_lines() {
        scan(
            """package abc
               import x
               class A {
                 fun foo(a: Int, b: Int) {
                   statementOnSeveralLines(a,
                     b)
                 }
               }
               fun bar() {
                 val x = 42
               }""")
        assertThat(visitor.executableLines()).containsExactly(5, 10)
    }

    private fun scan(code: String) {
        inputFile = TestInputFileBuilder("moduleKey", createTempFile(tempFolder).name)
            .setCharset(StandardCharsets.UTF_8)
            .initMetadata(code).build()
        val ctx = InputFileContextImpl(sensorContext, inputFile, false)
        visitor.scan(ctx, kotlinTreeOf(code, environment, inputFile))
    }
}
